src/webgpu/api/validation/queue/copyImageBitmapToTexture.spec.ts

```diff
@@ -0,0 +1,409 @@
+      .combine(poptions('sampleCount', [1, 4]))
+      .combine(poptions('mipLevel', [0, 2]))
+      .combine(poptions('format', kAllTextureFormats))
+  )
+  .fn(async t => {
+    const { destroyed, dimension, usage, sampleCount, mipLevel, format } = t.params;
+    const src = await t.getDefaultImageBitmapCopyView();
+    const dst = t.getTextureCopyViewWithDefaultSize(
+      sampleCount,
+      dimension,
+      format,
+      usage,
+      mipLevel,
+      { x: 0, y: 0, z: 0 },
+      destroyed
+    );
+    const copySize = { width: 1, height: 1, depth: 1 };
+
+    let success = true;
+    let exceptionName = '';
+
+    if (destroyed) {
+      success = false;
+    }
+
+    if (!(usage & GPUTextureUsage.COPY_DST)) {
+      // Usage must have GPUTextureUsage.COPY_DST
+      success = false;
+    }
+
+    if (sampleCount > 1) {
+      // Dst texture cannot be multisampled due to imageBitmap.
+      success = false;
+    }
+
+    if (mipLevel > 1) {
+      // MipLevel value cannot exceed dst texture max mipLevel, which is 1 in tests.
+      success = false;
+    }
+
+    if (!kValidTextureFormatsForCopyIB2T.includes(format)) {
+      // Dst texture format needs to be supported in copyImageBitmapToTexture.
+      success = false;
+      exceptionName = 'TypeError';
+    }
+
+    t.testRun(src, dst, copySize, success, exceptionName);
+  });
+
+g.test('copyImageBitmapToTexture_copy_size')
+  .params(
+    params()
+      .combine(poptions('dimension', ['2d'] as const))
+      .expand(copySizeTestForDstTextureSize)
+      .combine(
+        poptions('srcImageBitmapSize', [
+          { width: DEFAULT_WIDTH, height: DEFAULT_HEIGHT }, // case: src size <= dst size
+          { width: DEFAULT_WIDTH + 1, height: DEFAULT_HEIGHT + 1 }, // case: src size >= dst size
+        ])
+      )
+      .combine(poptions('srcOriginValue', generateSrcOriginValue()))
+      .combine(poptions('dstOriginValue', generateDstOriginValue()))
+      .combine(poptions('copySize', generateCopySizeValue()))
+  )
+  .fn(async t => {
+    const {
+      dimension,
+      dstTextureSize,
+      srcImageBitmapSize,
+      srcOriginValue,
+      dstOriginValue,
+      copySize,
+    } = t.params;
+
+    if (dimension !== '2d') {
+      t.skip('Skip test case for now because no 1d, 3d texture supported');
+    }
+
+    const src = await t.getImageBitmapCopyView(
+      srcImageBitmapSize.width,
+      srcImageBitmapSize.height,
+      srcOriginValue,
+      false
+    );
+
+    const dst = t.getTextureCopyView(
+      dstTextureSize,
+      1,
+      dimension,
+      'rgba8unorm',
+      GPUTextureUsage.COPY_DST,
+      0,
+      dstOriginValue,
+      false
+    );
+
+    let success = true;
+    let exceptionName = '';
+
+    if (
+      srcImageBitmapSize.width <= srcOriginValue.x ||
+      srcImageBitmapSize.width - srcOriginValue.x < copySize.width ||
+      srcImageBitmapSize.height <= srcOriginValue.y ||
+      srcImageBitmapSize.height - srcOriginValue.y < copySize.height ||
```

Comment:
Note that these operations operate on `doubles`. So I think it would be valid to write:
```suggestion
      srcOriginValue.x + copySize.width > srcImageBitmapSize.width ||
      srcOriginValue.y + copySize.height > srcImageBitmapSize.height ||
```
which is much easier to read.

Reply:
Done